usr/opt/OSC200/usr/share/man/man1/lprm.1
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: lprm.1,v $ $Revision: 3.5.5.2 $ (DEC) $Date: 1993/09/30 21:37:20 $
.\" 
...\" (c) Copyright 1990, 1991, 1992 OPEN SOFTWARE FOUNDATION, INC.
...\" ALL RIGHTS RESERVED
...\" OSF/1 Release 1.0.4
...\"build rmifdef
.TH lprm 1
.SH NAME
.PP
\*Llprm\*O \- Removes requests from the line printer spooling queue
.SH SYNOPSIS
.PP
.sS
\*Llprm\*O
\*O[\*L\-P\*Vprinter\*O]
\*O[\*L\-\*O]
\*O[\*Vrequest_ID\*O
\&.\&.\&.]
\*O[\*Vuser\*O
\&.\&.\&.]
.sE
.PP
The \*Llprm\*O command
removes one or more requests from a printer's spool queue.
.SH DESCRIPTION
.PP
.iX "lprm command"
.iX "canceling" "printer requests"
.iX "queues" "canceling requests"
.iX "printers" "canceling queued requests"
.iX "line printers" "spooling queue"
.iX "requests" "removing from printer queue"
Because the spooling directory is protected from users, using
.gL "spooling"
\*Llprm\*O is normally the way a user can remove a request.
.PP
The \*Llprm\*O command without any arguments
deletes the currently active request if it is
owned by the user who invoked \*Llprm\*O.
.PP
The \*L\-\*O flag removes all requests from the user issuing the command.
If a user who has
superuser authority
uses this flag, the spool queue is emptied
entirely.  The owner is determined by the user's username
and hostname on the machine where the \*Llpr\*O command was invoked.
.PP
You can remove an individual request from a queue by specifying its
request ID.  (You can obtain the request ID by using the \*Llpq\*O or
\*Llpstat\*O commands.)  Specifying one or more \*Vuser\*Os
removes any requests queued belonging to those \*Vuser\*Os.
This
works only for a user with superuser authority.
.PP
The \*Llprm\*O command is silent if there are no requests in
the queue that match the request list.
.PP
The \*Llprm\*O command kills an active daemon, if necessary, before
removing any spooling files.  If a daemon is killed, a new one is
automatically restarted upon completion of file removals.
.SH FLAGS
.PP
.VL 4m
.LI "\*L\-\*O"
Removes all requests that a user
owns.  If specified by a user
with superuser authority,
the spool queue is emptied entirely.
.LI "\*L\-P\*Vprinter\*O"
Specifies the queue associated with a specific
printer; otherwise, the default printer or the value of the \*LPRINTER\*O
variable in the environment is used.
.LE
.SH EXAMPLES
.PP
.AL
.LI
To remove a request from the default queue by specifying the request ID, enter:
.iS
lprm 13
\*O\c
.iE
.LI
To remove a request from a specific queue, enter:
.iS
lprm \-Plp0 13
\*O\c
.iE
.LI
To remove from a specific queue a request from user \*Lguest\*O, enter:
.iS
lprm \-Plp0 guest
\*O\c
.iE
.LE
.SH NOTES
.PP
Because race conditions are possible in the update of the lock
file, the currently active request may be incorrectly identified.
.SH DIAGNOSTICS
.PP
.VL 14m
.LI "\*CPermission denied\*O"
The user tried to remove files that belong to another user.
.LE
.SH FILES
.PP
.VL 11m
.LI "\*L/etc/printcap\*O"
Printer description file.
.LI "\*L/var/spool/*\*O"
Spool directories.
.LI "\*L/var/spool/*/cf*\*O"
Daemon control files.
.LI "\*L/var/spool/*/df*\*O"
Data files specified in \*Lcf\*O files.
.LI "\*L/var/spool/*/tf*\*O"
Temporary copies of \*Lcf\*O files.
.LI "\*L/var/spool/*/lock\*O"
Lock file used to obtain the process ID of the current
daemon and the request ID of the currently active request.
.LE
.SH RELATED INFORMATION
.PP
Commands: \*Llp\*O(1)/\*Lcancel\*O(1), \*Llpc\*O(8), \*Llpd\*O(8),
\*Llpq\*O(1), \*Llpr\*O(1), \*Llpstat\*O(1).
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/sbin/lpd/lprm.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: lprm.c,v $ $Revision: 4.2.6.2 $ (DEC) $Date: 1993/10/08 15:14:41 $";
#endif 
/*
 */
/*
 * (c) Copyright 1990, OPEN SOFTWARE FOUNDATION, INC.
 * ALL RIGHTS RESERVED
 */
/*
 * OSF/1 Release 1.0
 */
#if !defined(lint) && !defined(_NOIDENT)

#endif
/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice and this paragraph are
 * duplicated in all such forms and that any documentation,
 * advertising materials, and other materials related to such
 * distribution and use acknowledge that the software was developed
 * by the University of California, Berkeley.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 * 
 * lprm.c	5.4 (Berkeley) 6/30/88
 * lprm.c	4.1 15:58:45 7/19/90 SecureWare 
 */


/*
 * lprm - remove the current user's spool entry
 *
 * lprm [-] [[job #] [user] ...]
 *
 * Using information in the lock file, lprm will kill the
 * currently active daemon (if necessary), remove the associated files,
 * and startup a new daemon.  Priviledged users may remove anyone's spool
 * entries, otherwise one can only remove their own.
 */

#include "lp.h"
#include <locale.h>

#if SEC_BASE
#define MSGSTR_SEC(n,s) catgets(catd,MS_PRINTER_SEC,n,s)

#include <sys/security.h>

extern priv_t *privvec();
#endif

#if !SEC_BASE
extern	int notopr;	/* imported from rmjob */
#endif

/*
 * Stuff for handling job specifications
 */
char	*user[MAXUSERS];	/* users to process */
int	users;			/* # of users in user array */
int	requ[MAXREQUESTS];	/* job number of spool entries */
int	requests;		/* # of spool requests */
char	*person;		/* name of person doing lprm */

static char	luser[16];	/* buffer for person */

/* struct passwd *getpwuid(); */

main(argc, argv)
	char *argv[];
{
	register char *arg;
	struct passwd *p;
	struct direct **files;
	int nitems, assasinated = 0;
	void usage();

        (void) setlocale( LC_ALL, "" );
        catd = catopen(MF_PRINTER,NL_CAT_LOCALE);
#if SEC_BASE
	set_auth_parameters(argc, argv);
	initprivs();
#endif
	name = argv[0];
	gethostname(host, sizeof(host));
	openlog("lpd", 0, LOG_LPR);
#if SEC_BASE
	if ((p = getpwuid(getluid())) == NULL)
#else
	if ((p = getpwuid(getuid())) == NULL)
#endif
		fatal(MSGSTR(LPRM_1, "Who are you?"));
	if (strlen(p->pw_name) >= sizeof(luser))
		fatal(MSGSTR(LPRM_2, "Your name is too long"));
	strcpy(luser, p->pw_name);
	person = luser;
	while (--argc) {
		if ((arg = *++argv)[0] == '-')
			switch (arg[1]) {
			case 'P':
				if (arg[2])
					printer = &arg[2];
				else if (argc > 1) {
					argc--;
					printer = *++argv;
				}
				break;
			case '\0':
				if (!users) {
					users = -1;
					break;
				}
			default:
				usage();
			}
		else {
			if (users < 0)
				usage();
			if (isdigit(arg[0])) {
				if (requests >= MAXREQUESTS)
					fatal(MSGSTR(LPRM_4, "Too many requests"));
				requ[requests++] = atoi(arg);
			} else {
				if (users >= MAXUSERS)
					fatal(MSGSTR(LPRM_5, "Too many users"));
				user[users++] = arg;
			}
		}
	}
	if (printer == NULL && (printer = getenv("PRINTER")) == NULL)
		printer = DEFLP;

#if SEC_BASE
	/*
	 * Raise privileges needed by rmjob to kill the current daemon
	 * (in case we are removing the currently printing job) and to
	 * connect to a remote daemon in case the printer is remote.
	 * On systems with MAC configured, raise the override privilege
	 * so we can manipulate the queue regardless of our current SL.
	 */
	if (forceprivs(privvec(SEC_KILL, SEC_REMOTE,
#if SEC_MAC
				SEC_ALLOWMACACCESS,
#endif
				-1), (priv_t *) 0)) {
		fprintf(stderr, MSGSTR_SEC(PRIVS, "%s: insufficient privileges\n"), "lprm");
		exit(1);
	}
#endif
	rmjob();
}

static void usage()
{
	fprintf(stderr, MSGSTR(LPRM_6, "usage: lprm [-] [-Pprinter] [[job #] [user] ...]\n"));
	exit(2);
}

